### DISCUSSION TITLE
Need help annotating useHydrateAtoms parameter for testing

### DISCUSSION THREAD

**Discussion by ronaldcurtis** [pmndrs/jotai#1921] [Q&A] [answered]
Hi there,

I'm trying to use injected values in jest tests using the [documentation here](https://jotai.org/docs/guides/testing#injected-values).

But I'm having trouble annotating types correctly in Typescript. 

I believe it's related to [this PR ](https://github.com/pmndrs/jotai/pull/1898). cc @ericdowell 

My current workaround is:

```ts
import { useHydrateAtoms } from 'jotai/utils'
import { Provider } from 'jotai'
import { ReactElement } from 'react'
import { WritableAtom } from 'jotai/vanilla'

/**
 * Define these types as Jotai does not export them
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyWritableAtom = WritableAtom<unknown, any[], any>
type AtomTuple<A = AnyWritableAtom, V = unknown> = readonly [A, V]

function HydrateAtoms<T extends Array<AtomTuple>>({ initialValues, children }: JotaiTestProviderProps<T>) {
  useHydrateAtoms(initialValues)
  return children
}

type JotaiTestProviderProps<T extends Array<AtomTuple>> = {
  /**
   * useHydrateAtoms is overloaded, and I'm not sure how to annotate all the
   * type constraints, so we disable the following line from eslint and the type-checker.
   */
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  initialValues: Parameters<typeof useHydrateAtoms<T>>[0]
  children: ReactElement
}

export function JotaiTestProvider<T extends Array<AtomTuple>>({ initialValues, children }: JotaiTestProviderProps<T>) {
  return (
    <Provider>
      <HydrateAtoms initialValues={initialValues}>{children}</HydrateAtoms>
    </Provider>
  )
}

```

If I don't use `@ts-ignore` above, I get the following error:

<img width="1415" alt="Screen Shot 2023-05-08 at 9 30 59 am" src="https://user-images.githubusercontent.com/1256477/236707956-6efb8c67-86bf-4c8e-b652-99f29ec9ce8a.png">

Any help appreciated 🙏

**Comment by ericdowell** [comment]
I've opened a PR to add a `HydrateAtoms` component to Jotai in PR https://github.com/pmndrs/jotai/pull/1922. We'll see if the PR gets accepted or not.

**Reply by dai-shi** [reply]
Well, that's not acceptable in `jotai` package.
However, it it were `jotai-testing` package, it probably would.
```js
// imaginary code
import { TestProvider } from 'jotai-testing';
```

**Reply by ericdowell** [reply]
If you'd like to start supporting that I could see this working. Then it may be easier for developers to get started and docs can suggest using `jotai-testing` package.

**Reply by dai-shi** [reply]
I've wanted a dedicated package for testing if that helps. One thing people requested is to test atom's internal logic without using React https://github.com/pmndrs/jotai/issues/252#issuecomment-768266877. And also resetting stores.

I don't have the bandwidth to work on it, but if you or anyone else is interested, I would be happy to help set up a repository.

**Comment by dai-shi** [comment]
Maybe we should stop suggesting the use of HydrateAtoms in docs, if that causes too much trouble?

**Reply by dai-shi** [reply]
The problem seems harder than I thought...

**Reply by ericdowell** [reply]
Likely you'll want to update the [docs](https://jotai.org/docs/guides/testing#injected-values) to show working typescript version of `HydrateAtoms` so it's clear for future developers.

**Reply by dai-shi** [reply]
Needs a PR!

**Comment by ericdowell** [comment, marked as answer]
Looking more closely there's no need to recreate or pass `T` generics, so we can drop all of the `<T extends Array<AtomTuple>>` usages and no need to pass `T` to `useHydrateAtoms`.

Try using this:
```ts
import { ReactElement } from 'react'
import { Provider } from 'jotai'
import { useHydrateAtoms } from 'jotai/utils'

function HydrateAtoms({ initialValues, children }: JotaiTestProviderProps) {
  useHydrateAtoms(initialValues)
  return children
}

type JotaiTestProviderProps = {
  initialValues: Parameters<typeof useHydrateAtoms>[0]
  children: ReactElement
}

export function JotaiTestProvider({
  initialValues,
  children,
}: JotaiTestProviderProps) {
  return (
    <Provider>
      <HydrateAtoms initialValues={initialValues}>{children}</HydrateAtoms>
    </Provider>
  )
}
```

**Reply by ericdowell** [reply]
If the above works please mark the above comment as the answer since I've closed the PR https://github.com/pmndrs/jotai/pull/1922#event-9186082614 after further discussion with @dai-shi

**Reply by ronaldcurtis** [reply]
I tried the above as well prior to this, but when passing `initialValues` the type-checker complains about not matching the `never` type. I'll try and get a screenshot.

**Reply by ronaldcurtis** [reply]
<img width="1227" alt="Screen Shot 2023-05-08 at 2 22 02 pm" src="https://user-images.githubusercontent.com/1256477/236733268-f7e6c3e0-ebc9-4313-9097-d16271dcff85.png">

**Reply by ericdowell** [reply]
I wasn't able to make this work with an array.

But I was able to figure our something with `Map`. To keep things type safe I would suggest implementing a custom class which uses a `Map` under the hood, since `AtomMap` function overload type is generalized.

Feel free to change the method names of the `JotaiInitialValues` class. Otherwise, by using the `set` method in my example you get the type safety when passing the `values` for specific the `atom` just as `useHydrateAtoms` does normally with an array of arrays. The `set` method returns `this` which will allow chaining multiple `set` calls together.
```tsx
import { ReactElement } from 'react'
import { render } from '@testing-library/react'
import { it } from 'vitest'
import { Provider, WritableAtom, atom } from 'jotai'
import { useHydrateAtoms } from 'jotai/utils'

type AnyWritableAtom = WritableAtom<unknown, any[], any>

class JotaiInitialValues {
  protected atomMap: Map<AnyWritableAtom, unknown>

  constructor() {
    this.atomMap = new Map<AnyWritableAtom, unknown>()
  }

  set<T extends AnyWritableAtom>(atom: T, values: ReturnType<T['read']>) {
    this.atomMap.set(atom, values)
    return this
  }

  atoms() {
    return this.atomMap
  }
}

type JotaiTestProviderProps = {
  initialValues: JotaiInitialValues
  children: ReactElement
}

function HydrateAtoms({ initialValues, children }: JotaiTestProviderProps) {
  useHydrateAtoms(initialValues.atoms())
  return children
}

export function JotaiTestProvider({
  initialValues,
  children,
}: JotaiTestProviderProps) {
  return (
    <Provider>
      <HydrateAtoms initialValues={initialValues}>{children}</HydrateAtoms>
    </Provider>
  )
}

it('allows passing initialValues', () => {
  const complexCountAtom = atom({
    count: 0,
  })
  const activeAtom = atom(true)
  render(
    <JotaiTestProvider
      initialValues={new JotaiInitialValues()
        .set(complexCountAtom, {
          count: 1,
        })
        .set(activeAtom, false)}>
      <h1>Foo Bar Count</h1>
    </JotaiTestProvider>
  )
})
```
If you don't like using `new` keyword everywhere in tests just make a helper function:
```tsx
function jotaiTestValues() {
  return new JotaiInitialValues()
}

it('allows passing initialValues', () => {
  const complexCountAtom = atom({
    count: 0,
  })
  const activeAtom = atom(true)
  render(
    <JotaiTestProvider
      initialValues={jotaiTestValues()
        .set(complexCountAtom, {
          count: 1,
        })
        .set(activeAtom, false)}>
      <h1>Foo Bar Count</h1>
    </JotaiTestProvider>
  )
})
```

**Reply by ronaldcurtis** [reply]
@ericdowell Thanks heaps for that! Works like a charm 🎉
